File: .backup/src/router/path-discovery.ts
// ─────────────────────────────────────────────────────────────
// Path Discovery
// Finds candidate paths between source and destination chains
// ─────────────────────────────────────────────────────────────

import type { Chain, CandidatePath, Token, BridgeQuote } from '../types/index.js';
import { ALL_CHAINS } from '../types/index.js';
import type { BridgeAdapter } from '../bridges/adapter.js';
import { BridgeRegistry } from '../bridges/adapter.js';
import { findToken } from '../chains/index.js';

export interface PathDiscoveryOptions {
  maxHops: number;
  excludeBridges: string[];
  excludeChains: Chain[];
  minLiquidity: number;
}

/**
 * Chain connectivity map.
 * Defines which chains have strong bridge connectivity to each other.
 * This is used for heuristic path ordering, not hard filtering.
 */
export const CHAIN_CONNECTIVITY: Record<Chain, Chain[]> = {
  ethereum: ['arbitrum', 'optimism', 'base', 'polygon', 'bnb', 'avalanche', 'solana'],
  solana: ['ethereum', 'polygon', 'bnb', 'avalanche'],
  arbitrum: ['ethereum', 'optimism', 'base', 'polygon'],
  base: ['ethereum', 'arbitrum', 'optimism', 'polygon'],
  polygon: ['ethereum', 'arbitrum', 'bnb', 'avalanche', 'solana'],
  bnb: ['ethereum', 'polygon', 'avalanche', 'solana'],
  optimism: ['ethereum', 'arbitrum', 'base', 'polygon'],
  avalanche: ['ethereum', 'polygon', 'bnb'],
};

/**
 * Preferred intermediate tokens per chain.
 * These are the most liquid tokens available for bridging on each chain.
 */
export const INTERMEDIATE_TOKENS: Record<Chain, string[]> = {
  ethereum: ['USDC', 'USDT', 'WETH', 'DAI'],
  solana: ['USDC', 'USDT', 'SOL'],
  arbitrum: ['USDC', 'USDT', 'WETH'],
  base: ['USDC', 'WETH'],
  polygon: ['USDC', 'USDT', 'WETH'],
  bnb: ['USDC', 'USDT', 'BNB'],
  optimism: ['USDC', 'WETH', 'OP'],
  avalanche: ['USDC', 'USDT', 'AVAX'],
};

/**
 * Build an adjacency map of chain -> chain -> bridge adapters.
 */
function buildGraph(
  registry: BridgeRegistry,
  excludeBridges: string[],
): Map<Chain, Map<Chain, BridgeAdapter[]>> {
  const graph = new Map<Chain, Map<Chain, BridgeAdapter[]>>();
  for (const adapter of registry.getAll()) {
    if (excludeBridges.includes(adapter.name)) continue;
    const chains = adapter.supportedChains;
    for (const from of chains) {
      for (const to of chains) {
        if (from === to) continue;
        if (!adapter.supportsRoute(from, to)) continue;
        if (!graph.has(from)) graph.set(from, new Map());
        const inner = graph.get(from)!;
        if (!inner.has(to)) inner.set(to, []);
        inner.get(to)!.push(adapter);
      }
    }
  }
  return graph;
}

/**
 * Discover all viable chain-level paths from source to destination.
 * Returns arrays of chain sequences, e.g. [ethereum, arbitrum, solana].
 * Uses DFS with cycle detection and hop limits.
 */
export function discoverChainPaths(
  fromChain: Chain,
  toChain: Chain,
  registry: BridgeRegistry,
  options: PathDiscoveryOptions,
): Chain[][] {
  if (fromChain === toChain) return [];

  const graph = buildGraph(registry, options.excludeBridges);
  const results: Chain[][] = [];
  const visited = new Set<Chain>();

  function dfs(current: Chain, path: Chain[]): void {
    // path includes fromChain, so maxHops+1 is the length limit
    if (path.length > options.maxHops + 1) return;
    if (current === toChain) {
      results.push([...path]);
      return;
    }
    const neighbors = graph.get(current);
    if (!neighbors) return;

    // Sort neighbors by connectivity heuristic: prefer chains that are
    // well-connected to the destination
    const neighborChains = Array.from(neighbors.keys());
    const sorted = sortByConnectivity(neighborChains, toChain);

    for (const next of sorted) {
      if (visited.has(next)) continue;
      if (options.excludeChains.includes(next) && next !== toChain) continue;
      visited.add(next);
      path.push(next);
      dfs(next, path);
      path.pop();
      visited.delete(next);
    }
  }

  visited.add(fromChain);
  dfs(fromChain, [fromChain]);
  return results;
}

/**
 * Sort chains by how well-connected they are to the target chain.
 * Chains directly connected to the target come first.
 */
function sortByConnectivity(chains: Chain[], target: Chain): Chain[] {
  return [...chains].sort((a, b) => {
    // Direct connection to target is best
    const aDirectly = a === target ? -100 : 0;
    const bDirectly = b === target ? -100 : 0;
    if (aDirectly !== bDirectly) return aDirectly - bDirectly;

    // Prefer chains connected to target
    const aConnected = CHAIN_CONNECTIVITY[a]?.includes(target) ? -10 : 0;
    const bConnected = CHAIN_CONNECTIVITY[b]?.includes(target) ? -10 : 0;
    if (aConnected !== bConnected) return aConnected - bConnected;

    // Prefer chains with more overall connectivity
    const aConns = CHAIN_CONNECTIVITY[a]?.length ?? 0;
    const bConns = CHAIN_CONNECTIVITY[b]?.length ?? 0;
    return bConns - aConns;
  });
}

/**
 * Filter dominated paths: remove paths that are strictly worse than another
 * (more hops through same intermediate chains).
 */
export function filterDominatedPaths(paths: Chain[][]): Chain[][] {
  const dominated = new Set<number>();
  for (let i = 0; i < paths.length; i++) {
    if (dominated.has(i)) continue;
    for (let j = 0; j < paths.length; j++) {
      if (i === j || dominated.has(j)) continue;
      if (paths[j].length > paths[i].length && isSubpath(paths[i], paths[j])) {
        dominated.add(j);
      }
    }
  }
  return paths.filter((_, idx) => !dominated.has(idx));
}

/**
 * Check if shorter path is a subsequence of longer path (same start/end).
 */
function isSubpath(shorter: Chain[], longer: Chain[]): boolean {
  if (shorter[0] !== longer[0]) return false;
  if (shorter[shorter.length - 1] !== longer[longer.length - 1]) return false;
  let si = 0;
  for (let li = 0; li < longer.length && si < shorter.length; li++) {
    if (longer[li] === shorter[si]) si++;
  }
  return si === shorter.length;
}

/**
 * Get the list of intermediate chains commonly used between two chains.
 * Useful for suggesting multi-hop paths.
 */
export function getIntermediateChains(from: Chain, to: Chain): Chain[] {
  const fromConns = new Set(CHAIN_CONNECTIVITY[from] ?? []);
  const toConns = new Set(CHAIN_CONNECTIVITY[to] ?? []);
  const intermediates: Chain[] = [];

  for (const chain of ALL_CHAINS) {
    if (chain === from || chain === to) continue;
    // A good intermediate is connected to both source and destination
    if (fromConns.has(chain) && toConns.has(chain)) {
      intermediates.push(chain);
    }
  }

  return intermediates;
}

/**
 * Get all bridge adapters that can handle a specific chain pair.
 */
export function getBridgesForPair(
  from: Chain,
  to: Chain,
  registry: BridgeRegistry,
  excludeBridges: string[] = [],
): BridgeAdapter[] {
  return registry
    .getForPair(from, to)
    .filter((a) => !excludeBridges.includes(a.name));
}

/**
 * Resolve a token for a chain. First tries the chain registry,
 * then falls back to constructing a placeholder.
 */
function resolveToken(chain: Chain, symbol: string): Token {
  const found = findToken(chain, symbol);
  if (found) return found;
  // Fallback: construct a token with common defaults
  const isStable = ['USDC', 'USDT', 'DAI'].includes(symbol.toUpperCase());
  return {
    symbol: symbol.toUpperCase(),
    chain,
    decimals: isStable ? 6 : 18,
    address: `0x${chain}_${symbol.toLowerCase()}`,
  };
}

/**
 * Select the best intermediate token for a hop between two chains.
 * Prefers stablecoins, then the same token as the source.
 */
function selectIntermediateToken(chain: Chain, preferredSymbol: string): Token {
  // If the preferred token exists on this chain, use it
  const preferred = findToken(chain, preferredSymbol);
  if (preferred) return preferred;

  // Otherwise, use USDC as the universal intermediate
  const usdc = findToken(chain, 'USDC');
  if (usdc) return usdc;

  // Fallback
  return resolveToken(chain, 'USDC');
}

/**
 * Compute the cartesian product of arrays.
 */
function cartesianProduct<T>(arrays: T[][]): T[][] {
  if (arrays.length === 0) return [[]];
  const [first, ...rest] = arrays;
  const restProduct = cartesianProduct(rest);
  const result: T[][] = [];
  for (const item of first) {
    for (const combo of restProduct) {
      result.push([item, ...combo]);
    }
  }
  return result;
}

/**
 * Limit the number of bridge combinations to prevent combinatorial explosion.
 * For each hop, keep only the top N bridges sorted by estimated output.
 */
function limitCombinations(
  adaptersPerHop: BridgeAdapter[][],
  maxPerHop: number = 4,
): BridgeAdapter[][] {
  return adaptersPerHop.map((adapters) => {
    if (adapters.length <= maxPerHop) return adapters;
    // Keep only the first N (they should already be in a reasonable order)
    return adapters.slice(0, maxPerHop);
  });
}

/**
 * Assign bridges and build candidate paths with quotes.
 * Fetches quotes from each bridge for each hop and constructs
 * fully-quoted candidate paths.
 */
export async function buildCandidatePaths(
  chainPaths: Chain[][],
  fromToken: Token,
  toToken: Token,
  amount: string,
  registry: BridgeRegistry,
  options: PathDiscoveryOptions,
): Promise<CandidatePath[]> {
  const candidates: CandidatePath[] = [];

  for (const chainPath of chainPaths) {
    const adaptersPerHop: BridgeAdapter[][] = [];
    let viable = true;

    for (let i = 0; i < chainPath.length - 1; i++) {
      const from = chainPath[i];
      const to = chainPath[i + 1];
      const bridgeAdapters = getBridgesForPair(from, to, registry, options.excludeBridges);
      if (bridgeAdapters.length === 0) {
        viable = false;
        break;
      }
      adaptersPerHop.push(bridgeAdapters);
    }

    if (!viable) continue;

    // Limit combinations to avoid exponential blowup
    const limited = limitCombinations(adaptersPerHop);
    const combinations = cartesianProduct(limited);

    for (const combo of combinations) {
      const bridgeNames = combo.map((a) => a.name);
      const tokens: Token[] = [fromToken];
      const quotes: BridgeQuote[] = [];
      let currentAmount = amount;
      let totalFee = 0;
      let totalTime = 0;
      let quotesFailed = false;

      for (let i = 0; i < combo.length; i++) {
        const hopFromToken = i === 0
          ? fromToken
          : selectIntermediateToken(chainPath[i], fromToken.symbol);
        const hopToToken = i === combo.length - 1
          ? toToken
          : selectIntermediateToken(chainPath[i + 1], fromToken.symbol);

        try {
          const quote = await combo[i].getQuote({
            fromChain: chainPath[i],
            toChain: chainPath[i + 1],
            fromToken: hopFromToken,
            toToken: hopToToken,
            amount: currentAmount,
            slippageTolerance: 50,
          });

          // Check minimum liquidity
          if (quote.liquidityDepth < options.minLiquidity) {
            quotesFailed = true;
            break;
          }

          quotes.push(quote);
          tokens.push(hopToToken);
          totalFee += parseFloat(quote.fee);
          totalTime += quote.estimatedTime;
          currentAmount = quote.outputAmount;
        } catch {
          quotesFailed = true;
          break;
        }
      }

      if (quotesFailed) continue;

      const outputAmount = parseFloat(currentAmount);
      const inputAmount = parseFloat(amount);
      const roughScore = inputAmount > 0 ? outputAmount / inputAmount : 0;

      candidates.push({
        chains: chainPath,
        bridges: bridgeNames,
        tokens,
        quotes,
        estimatedFee: totalFee,
        estimatedTime: totalTime,
        roughScore,
      });
    }
  }

  // Sort by rough score descending
  candidates.sort((a, b) => b.roughScore - a.roughScore);

  return candidates;
}

/**
 * PathDiscovery class - wraps the functional API for convenience.
 */
export class PathDiscovery {
  private registry: BridgeRegistry;
  private defaultOptions: PathDiscoveryOptions;

  constructor(registry: BridgeRegistry, options?: Partial<PathDiscoveryOptions>) {
    this.registry = registry;
    this.defaultOptions = {
      maxHops: options?.maxHops ?? 3,
      excludeBridges: options?.excludeBridges ?? [],
      excludeChains: options?.excludeChains ?? [],
      minLiquidity: options?.minLiquidity ?? 1000,
    };
  }

  /**
   * Discover all candidate paths for a request.
   */
  async discoverPaths(
    fromChain: Chain,
    toChain: Chain,
    fromToken: Token,
    toToken: Token,
    amount: string,
    options?: Partial<PathDiscoveryOptions>,
  ): Promise<CandidatePath[]> {
    const opts: PathDiscoveryOptions = { ...this.defaultOptions, ...options };

    // Step 1: Discover chain-level paths
    const chainPaths = discoverChainPaths(fromChain, toChain, this.registry, opts);

    if (chainPaths.length === 0) return [];

    // Step 2: Filter dominated paths
    const filtered = filterDominatedPaths(chainPaths);

    // Step 3: Build fully-quoted candidates
    const candidates = await buildCandidatePaths(
      filtered,
      fromToken,
      toToken,
      amount,
      this.registry,
      opts,
    );

    return candidates;
  }

  /**
   * Find only direct (single-hop) paths.
   */
  async expandDirectPaths(
    fromChain: Chain,
    toChain: Chain,
    fromToken: Token,
    toToken: Token,
    amount: string,
  ): Promise<CandidatePath[]> {
    return this.discoverPaths(fromChain, toChain, fromToken, toToken, amount, {
      maxHops: 1,
    });
  }

  /**
   * Find multi-hop paths (2+ hops).
   */
  async expandMultiHopPaths(
    fromChain: Chain,
    toChain: Chain,
    fromToken: Token,
    toToken: Token,
    amount: string,
    maxHops: number = 3,
  ): Promise<CandidatePath[]> {
    const all = await this.discoverPaths(
      fromChain, toChain, fromToken, toToken, amount,
      { maxHops },
    );
    return all.filter((c) => c.chains.length > 2);
  }
}

File: engine/src/path_discovery.rs
use crate::bridge::BridgeRegistry;
use crate::types::{Chain, Token};

/// Candidate path: a sequence of (from_chain, to_chain, bridge_name) tuples.
#[derive(Debug, Clone)]
pub struct CandidatePath {
    pub steps: Vec<PathStep>,
}

#[derive(Debug, Clone)]
pub struct PathStep {
    pub from_chain: Chain,
    pub to_chain: Chain,
    pub from_token: Token,
    pub to_token: Token,
    pub bridge_name: String,
}

/// Discovers all possible paths between two chain/token pairs.
pub struct PathDiscovery<'a> {
    registry: &'a BridgeRegistry,
    max_hops: usize,
}

impl<'a> PathDiscovery<'a> {
    pub fn new(registry: &'a BridgeRegistry, max_hops: usize) -> Self {
        Self { registry, max_hops }
    }

    /// Enumerate all candidate paths from source to destination.
    pub fn discover_paths(
        &self,
        from_chain: Chain,
        from_token: &Token,
        to_chain: Chain,
        to_token: &Token,
    ) -> Vec<CandidatePath> {
        let mut all_paths = Vec::new();

        // Direct paths (1 hop)
        let direct = self.expand_direct_paths(from_chain, from_token, to_chain, to_token);
        all_paths.extend(direct);

        // Multi-hop paths (2-3 hops)
        if self.max_hops >= 2 {
            let multi = self.expand_multi_hop_paths(from_chain, from_token, to_chain, to_token);
            all_paths.extend(multi);
        }

        // Remove dominated and duplicate paths
        let filtered = self.filter_dominated_paths(all_paths);
        self.deduplicate_paths(filtered)
    }

    /// Find all single-hop direct bridges between two chains.
    pub fn expand_direct_paths(
        &self,
        from_chain: Chain,
        from_token: &Token,
        to_chain: Chain,
        to_token: &Token,
    ) -> Vec<CandidatePath> {
        let bridges = self.registry.get_bridges_for_pair(from_chain, to_chain);
        bridges
            .into_iter()
            .map(|bridge| {
                CandidatePath {
                    steps: vec![PathStep {
                        from_chain,
                        to_chain,
                        from_token: from_token.clone(),
                        to_token: to_token.clone(),
                        bridge_name: bridge.name().to_string(),
                    }],
                }
            })
            .collect()
    }

    /// Expand multi-hop paths through intermediate chains.
    pub fn expand_multi_hop_paths(
        &self,
        from_chain: Chain,
        from_token: &Token,
        to_chain: Chain,
        to_token: &Token,
    ) -> Vec<CandidatePath> {
        let mut paths = Vec::new();
        let intermediates = self.get_intermediate_chains(from_chain, to_chain);

        // 2-hop paths: from -> intermediate -> to
        for &mid_chain in &intermediates {
            let mid_token = self.infer_intermediate_token(from_token, mid_chain);
            let first_leg = self.registry.get_bridges_for_pair(from_chain, mid_chain);
            let second_leg = self.registry.get_bridges_for_pair(mid_chain, to_chain);

            for bridge1 in &first_leg {
                for bridge2 in &second_leg {
                    paths.push(CandidatePath {
                        steps: vec![
                            PathStep {
                                from_chain,
                                to_chain: mid_chain,
                                from_token: from_token.clone(),
                                to_token: mid_token.clone(),
                                bridge_name: bridge1.name().to_string(),
                            },
                            PathStep {
                                from_chain: mid_chain,
                                to_chain,
                                from_token: mid_token.clone(),
                                to_token: to_token.clone(),
                                bridge_name: bridge2.name().to_string(),
                            },
                        ],
                    });
                }
            }
        }

        // 3-hop paths (only if max_hops >= 3)
        if self.max_hops >= 3 {
            for &mid1 in &intermediates {
                let mid1_token = self.infer_intermediate_token(from_token, mid1);
                let second_intermediates = self.get_intermediate_chains(mid1, to_chain);

                for &mid2 in &second_intermediates {
                    if mid2 == from_chain || mid2 == mid1 {
                        continue;
                    }
                    let mid2_token = self.infer_intermediate_token(from_token, mid2);

                    let leg1 = self.registry.get_bridges_for_pair(from_chain, mid1);
                    let leg2 = self.registry.get_bridges_for_pair(mid1, mid2);
                    let leg3 = self.registry.get_bridges_for_pair(mid2, to_chain);

                    if leg1.is_empty() || leg2.is_empty() || leg3.is_empty() {
                        continue;
                    }

                    // For 3-hop, only use the best bridge per leg to limit combinatorics
                    let b1_name = leg1[0].name().to_string();
                    let b2_name = leg2[0].name().to_string();
                    let b3_name = leg3[0].name().to_string();

                    paths.push(CandidatePath {
                        steps: vec![
                            PathStep {
                                from_chain,
                                to_chain: mid1,
                                from_token: from_token.clone(),
                                to_token: mid1_token.clone(),
                                bridge_name: b1_name,
                            },
                            PathStep {
                                from_chain: mid1,
                                to_chain: mid2,
                                from_token: mid1_token.clone(),
                                to_token: mid2_token.clone(),
                                bridge_name: b2_name,
                            },
                            PathStep {
                                from_chain: mid2,
                                to_chain,
                                from_token: mid2_token.clone(),
                                to_token: to_token.clone(),
                                bridge_name: b3_name,
                            },
                        ],
                    });
                }
            }
        }

        paths
    }

    /// Filter out paths that are provably dominated by another path.
    /// A path is dominated if another path uses a subset of its bridges
    /// with fewer hops (fewer hops generally means lower fees).
    pub fn filter_dominated_paths(&self, paths: Vec<CandidatePath>) -> Vec<CandidatePath> {
        if paths.len() <= 1 {
            return paths;
        }

        let mut kept = Vec::new();

        for (i, path) in paths.iter().enumerate() {
            let mut is_dominated = false;
            for (j, other) in paths.iter().enumerate() {
                if i == j {
                    continue;
                }
                // other dominates path if it has fewer hops and covers the same chain pair
                if other.steps.len() < path.steps.len() {
                    let same_endpoints = path.steps.first().map(|s| s.from_chain)
                        == other.steps.first().map(|s| s.from_chain)
                        && path.steps.last().map(|s| s.to_chain)
                            == other.steps.last().map(|s| s.to_chain);

                    let uses_same_bridge = other.steps.iter().any(|os| {
                        path.steps.iter().any(|ps| ps.bridge_name == os.bridge_name)
                    });

                    if same_endpoints && uses_same_bridge {
                        is_dominated = true;
                        break;
                    }
                }
            }
            if !is_dominated {
                kept.push(path.clone());
            }
        }

        kept
    }

    /// Remove duplicate paths that traverse the same chain sequence with the same bridges.
    pub fn deduplicate_paths(&self, paths: Vec<CandidatePath>) -> Vec<CandidatePath> {
        let mut seen = std::collections::HashSet::new();
        let mut unique = Vec::new();

        for path in paths {
            let key = path
                .steps
                .iter()
                .map(|s| {
                    format!(
                        "{}->{}:{}",
                        s.from_chain.chain_id(),
